x/interchainquery/keeper/msg_server.go

```diff
@@ -93,6 +94,34 @@ func (k Keeper) VerifyKeyProof(ctx sdk.Context, msg *types.MsgSubmitQueryRespons
 	return nil
 }
 
+// Checks if the query timed out, and if so, checks the timeout policy
+// Returns whether the callback should be executed (only occurs if explicitly indicated)
+func (k Keeper) CheckForQueryTimeout(ctx sdk.Context, query types.Query) (executeCallback bool, err error) {
+	if !query.HasTimedOut(ctx) {
+		return true, nil
+	}
+
+	k.Logger(ctx).Error(utils.LogICQCallbackWithHostZone(query.ChainId, query.CallbackId,
+		"QUERY TIMEOUT - QueryId: %s, TTL: %d, BlockTime: %d", query.Id, query.TimeoutTimestamp, ctx.BlockHeader().Time.UnixNano()))
+
+	switch query.TimeoutPolicy {
+	case types.TimeoutPolicy_REJECT_QUERY_RESPONSE:
+		k.Logger(ctx).Info(utils.LogICQCallbackWithHostZone(query.ChainId, query.CallbackId, "Rejecting query"))
+		return false, nil
+
+	case types.TimeoutPolicy_RETRY_QUERY_REQUEST:
+		k.Logger(ctx).Info(utils.LogICQCallbackWithHostZone(query.ChainId, query.CallbackId, "Retrying query..."))
+		return false, k.RetryICQRequest(ctx, query)
```

Comment:
So this function can actually fire off a query retry in this case... 
The "check" function was a pure functional computation except for this case which can have side effect so maybe we should name it handle timedout queries? If it is ok for this function to do the work then why return a boolean about if the callback should be executed instead of executing the callback right here?

Reply:
yeah I hear ya. The original motivation was so that we wouldn't have to call InvokeCallback two different times. But I see the confusion, so maybe it's best to remove this function or do the refactor above